### ISSUE TITLE
AsyncPipe doesn't update when changeDetectorRef.detach()

### ISSUE THREAD

**Issue by danishin** [angular/angular#12356]
**I'm submitting a ...**  (check one with "x")

```
[x] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
`AsyncPipe` doesn't update when `ChangeDetectorRef` for the given component is detached.

<hr>

Now, I'm not sure if this is a bug / feature (probably, very likely a feature).
By looking at [the way change detection is implemented](https://github.com/angular/angular/blob/master/modules/%40angular/common/src/pipes/async_pipe.ts#L137) in `AsyncPipe`, it seems like `this._ref.markForCheck()` doesn't run if `ChangeDetectorRef` is detached (as is also confirmed in [here](https://github.com/angular/angular/issues/7311#issuecomment-189292079)).

The motivation for this is to have full control over when the change detection occurs (by using `Observable` exclusively - alike MVVM pattern). I'm working a very performance-sensitive code where I have full control over inputs but using `ChangeDetectionStrategy.OnPush` fires change detection not just for input changes but also control event fires (like click events), which degrades the performance greatly by re-running unwanted change detection sweep on every click. 

So I opted to detach `ChangeDetectorRef` completely but then now the `AsyncPipe` stops working. Is this expected behavior for `AsyncPipe`? And if so, is there a way to work around this? Or perhaps, a way to make `ChangeDetectorStrategy.OnPush` to not fire on control event fires?

**Comment by DzmitryShylovich** [comment]
Can u provide a minimal working reproduction in plunkr?

**Comment by DzmitryShylovich** [comment]
@danishin What's the expected behavior? Should we reattach CDR call `markForCheck` and then detach it again?

**Comment by danishin** [comment]
@DzmitryShylovich Yes.
```
// ...
_ref.reattach();
_ref.markForCheck();
_ref.detach();
// ...
```
would work.

**Comment by DzmitryShylovich** [comment]
@danishin as u can see in the plunker http://plnkr.co/edit/CyM5x8t9gSqYtoP1v0yP?p=preview async pipe doesn't receive new values in detached mode so everything is more complicated than we thought :)

**Comment by igmdvc** [comment]
Any updates on this? I saw a fix proposal https://github.com/angular/angular/issues/12416 that was closed with no explanation. Stateful pipes are horrible for performance so it would nice if we can control changes, for example, for one time binding in which the value comes from an async event, or just update a value when the observable emits a new value, this latter being the best approach IMHO.

**Comment by DzmitryShylovich** [comment]
@igmdvc don't use async pipe. subscribe inside a component and call `detectChanges()` yourself

**Comment by igmdvc** [comment]
@DzmitryShylovich your solution should work but make things a little painful and verbose for the case I want to implement, in which I need a pipe for translations in which we retrieve the translation from HTTP, so we would like to use something like:
`return new AsyncPipe(_ref).transform(callToServiceWhichReturnsAnObservable(value))`

I would like to skip the component here as it is not responsible for that logic and give the responsibility to the 'translate' pipe. It would be ideal to deal with async operation not only at component level but also at pipe level without going stateful.

**Comment by DzmitryShylovich** [comment]
@igmdvc can u provide a small repro in plunker of what are u trying to accomplish?

**Comment by sandangel** [comment]
May be this help?
https://stackblitz.com/edit/angular-blbxac

**Comment by evertonrobertoauler** [comment]
Thanks @sandangel, your demo shows how to currently work around this issue, but I'm super glad that this is been fixed with the new Ivy render and Change Detection that will start work without the need of zones.js and will trigger change detection automatically in this cases where zone.js is really unnecessary. 
https://blog.angularindepth.com/angular-ivy-change-detection-execution-are-you-prepared-ab68d4231f2c

**Comment by pkozlowski-opensource** [comment]
OK, so I feel like there is some confusion going on here regarding how change detection and detaching change detector works. Let's take https://stackblitz.com/edit/angular-kdvl3j example from @evertonrobertoauler and dig into what is going on in there.

First of all, let state that in the current Angular calling `detach()` on `ChangeDetectorRef` means: don't run change detection (check bindings, process child components etc.) in a _template_ of a given component. In practice this means that _data_ might change but the affected bindings will _not_ be re-evaluated (this is what change detection does, after all...).

In the example stackblitz an async pipe is used in a regular binding (`{{ number$ | async }}`). If you detach change detector for this component template you will never evaluate this binding and an async pipe will never run. 

Now, it _seems_ like this issue is about _detaching_ from change detector but _still_ having change detection running... We can't have both - either you want Angular to run change detection for you or not. As soon as one detaches change detection you are saying "hey, I'm in control here, I know when to run CD". So, to make things work you need to invoke change detection yourself, as proposed by @sandangel . I would just update @sandangel proposal with one important remark - you do _not_ need to re-attach  a view to run a change detection on it (!) - this is totally unnecessary. Here is an updated stackblitz: https://stackblitz.com/edit/angular-rzvkdx?file=src%2Fapp%2Fapp.component.ts

I'm going to close this issue here as it is mostly about the confusion on how the current system works and I hope that my clarifications above are sufficient to get people going / avoid those pitfalls in the future.

**Comment by lppedd** [comment]
@pkozlowski-opensource If I decide to detach() my "root" components, and thus manually invoke change detection as described, wil this still hold true with Ivy? What I mean is, am I now writing code that will be useless in Ivy?
